src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs

```diff
@@ -351,8 +356,37 @@ public static bool TryText(FormattingInfo formatInfo, IRContext irContext, Formu
                 return false;
             }
 
-            Contract.Assert(StringValue.AllowedListConvertToString.Contains(value.Type));
-
+            Contract.Assert(StringValue.AllowedListConvertToString.Contains(value.Type));
+
+            bool hasFormatCulture = false;
+            string enUsCultureName = "en-US";
+            CultureInfo formatCulture = culture;
+            string revFormatStr = formatString;
+
+            if (!string.IsNullOrEmpty(formatString) && formatString.StartsWith("[$-", StringComparison.OrdinalIgnoreCase))
+            {
+                int startIdx = formatString.IndexOf("[$-", StringComparison.Ordinal) + 3;
+                int endIdx = formatString.IndexOf("]", StringComparison.Ordinal);
+
+                if (endIdx > 0)
+                {
+                    hasFormatCulture = true;
+                    string formatCultureName = formatString.Substring(startIdx, endIdx - startIdx);
+
+                    if (!TryGetCulture(formatCultureName, out formatCulture))
+                    {
+                        result = new StringValue(irContext, formatCultureName);
+                        return false;
+                    }
+
+                    formatString = formatString.Substring(endIdx + 1);
+                    revFormatStr = formatString.Replace(",", "!");
+                    revFormatStr = revFormatStr.Replace(".", ",");
+                    revFormatStr = revFormatStr.Replace(" ", ",");
+                    revFormatStr = revFormatStr.Replace("!", ".");

```

Comment:
`'!'` and `' '` are valid characters in the format string (are output "as-is"). Something like `Text(1234.5678,"0 # # # !!!")` is valid (and outputs "1 2 3 5 !!!"

Reply:
update here